### PR TITLE
Add TEFusedDenseMLP for Dense+Grouped GEMM fusion on SM100+ (#4318)

### PR DESCRIPTION
**Pull request by sraman-rgb**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/extensions/transformer_engine.py [hunk 1]
```diff
@@ -2557,8 +2557,193 @@ def as_mlp_submodule(
                 ffn_hidden_size=ffn_hidden_size,
             )
 
+    class TEFusedDenseMLP(TEFusedMLP):
```
Comment (timmoon10) [inline megatron/core/extensions/transformer_engine.py]: This is not the best name since `TEFusedMLP` is already a fused dense MLP. Really this is a hack to reuse the grouped GEMM kernels, so maybe something like `TEFusedMLPWithGroupedLinear` would be more descriptive. Alternatively, this logic could be added to `TEFusedMLP._make_fused_impl` without creating a new class.

File: megatron/core/extensions/transformer_engine.py [hunk 2]
```diff
@@ -2557,8 +2557,193 @@ def as_mlp_submodule(
...
+        def _make_fused_impl(self) -> te.pytorch.ops.Sequential:
+            """Construct fused module with GroupedLinear(num_groups=1) + ScaledSwiGLU."""
+
+            fused_impl = te.pytorch.ops.Sequential()
+
+            # Tensor parallelism configuration
+            tp_world_size = get_tensor_model_parallel_world_size()
+            tp_group = None
+            if tp_world_size > 1:
+                tp_group = get_tensor_model_parallel_group()
```
Comment (timmoon10) [inline megatron/core/extensions/transformer_engine.py]: Grouped linear doesn't support tensor parallelism. If tensor parallelism is enabled, we should either error out or fallback to the impl with dense linear ops.

File: megatron/core/extensions/transformer_engine.py [hunk 3]
```diff
@@ -2557,8 +2557,193 @@ def as_mlp_submodule(
...
+                op.weight = self.linear_fc1.layer_norm_weight
+                op.bias = self.linear_fc1.layer_norm_bias
+            elif norm_type == "RMSNorm":
+                op = te.pytorch.ops.RMSNorm(norm_shape, **kwargs)
+                op.weight = self.linear_fc1.layer_norm_weight
+            else:
+                raise ValueError(f"Unsupported normalization ({norm_type})")
+            # Store norm in a separate Sequential applied OUTSIDE the MXFP8 autocast
+            # in forward(). Running norm inside MXFP8 context corrupts the saved rstd
+            # used in RMSNorm backward, causing gradient amplification up to 10^6.
```
Comment (timmoon10) [inline megatron/core/extensions/transformer_engine.py:2640]: This is a worrying bug. Do we know if it's from TE or cuDNN?

File: megatron/core/extensions/transformer_engine.py [hunk 4]
```diff
@@ -2557,8 +2557,193 @@ def as_mlp_submodule(
...
+            # Wrapped in tuple to avoid nn.Module submodule registration (which would
+            # duplicate the shared norm weight in state_dict/parameters).
+            norm_seq = te.pytorch.ops.Sequential()
+            norm_seq.append(op)
+            self._norm_seq = (norm_seq,)
+
+            # GLU interleave size must match ScaledSwiGLU and the CuTe kernel.
+            _GLU_INTERLEAVE_SIZE = 32
```
Comment (timmoon10) [inline megatron/core/extensions/transformer_engine.py:2648]: GLU interleaving should be configured in the base `MLP` module so checkpointing is handled correctly.

File: megatron/core/extensions/transformer_engine.py [hunk 5]
```diff
@@ -2557,8 +2557,193 @@ def as_mlp_submodule(
...
+            # Both are created once and reused — avoids object creation every call.
+            if not hasattr(self, '_recipe'):
+                if os.getenv("FP4_RECIPE", "") == "nvfp4":
+                    self._recipe = te.common.recipe.NVFP4BlockScaling()
+                else:
+                    self._recipe = te.common.recipe.MXFP8BlockScaling()
+            recipe = self._recipe
+
+            if self._fused_impl is None:
+                with te.pytorch.fp8_autocast(enabled=True, fp8_recipe=recipe):
```
Comment (timmoon10) [inline megatron/core/extensions/transformer_engine.py]: We are only constructing modules and not performing a forward pass, so we should use the model-init context instead:
```suggestion
                with te.pytorch.quantized_model_init(enabled=True, recipe=recipe):
```

File: megatron/core/extensions/transformer_engine.py [hunk 6]
```diff
@@ -2557,8 +2557,193 @@ def as_mlp_submodule(
...
+                    self._fused_impl = (self._make_fused_impl(),)
+
+            # Apply norm in BF16 OUTSIDE the MXFP8 autocast to preserve the rstd
+            # tensor used by RMSNorm backward (running it inside causes up to 10^6
+            # gradient amplification, and causes convergence issues).
+            normed = self._norm_seq[0](hidden_states_2d)
+
+            with te.pytorch.fp8_autocast(enabled=True, fp8_recipe=recipe):
```
Comment (timmoon10) [inline megatron/core/extensions/transformer_engine.py]: `fp8_autocast` has been deprecated since we've moved past FP8:
```suggestion
            with te.pytorch.autocast(enabled=True, recipe=recipe):
```

**Review by santhnm2 — APPROVED** [review]

File: megatron/core/transformer/transformer_config.py [hunk 7]
```diff
@@ -746,6 +746,12 @@ class TransformerConfig(ModelParallelConfig):
     If negative, generates bias once per layer and reuses it (abs value is std).
     This is an experimental feature for benchmarking purposes."""
 
+    dense_grouped_gemm: bool = False
```
Comment (timmoon10) [inline megatron/core/transformer/transformer_config.py]: This name is not clear to me. Something like the following would be clearer:
```suggestion
    use_grouped_gemm_for_dense_mlp: bool = False
```

File: megatron/core/extensions/transformer_engine.py (hunk 1 above)
Comment (sraman-rgb) [reply megatron/core/extensions/transformer_engine.py]: changed in the latest commit

File: megatron/core/extensions/transformer_engine.py (hunk 2 above)
Comment (sraman-rgb) [reply megatron/core/extensions/transformer_engine.py]: changed in the latest commit

File: megatron/core/extensions/transformer_engine.py (hunk 5 above)
Comment (sraman-rgb) [reply megatron/core/extensions/transformer_engine.py]: changed in the latest commit

File: megatron/core/extensions/transformer_engine.py (hunk 4 above)
Comment (sraman-rgb) [reply megatron/core/extensions/transformer_engine.py:2648]: currently, there is a separate swizzling that we do in checkpointing.py in MBridge to make sure the weights are loaded in interleaved fashion. Hence, I didnt change anything over here

File: megatron/core/transformer/transformer_config.py (hunk 7 above)
Comment (sraman-rgb) [reply megatron/core/transformer/transformer_config.py]: changed in the latest commit

**Review by ericharper — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
